### PR TITLE
test : getNetworkID method

### PR DESCRIPTION
**Pull request by KingDaemonX**
**What this PR does / why we need it**:

**Which issue this PR fixes(if applicable)**:
fixes #

**Special notes for reviewers**:

**Release note**:

```release-note
NONE
```
test for the important and complex logic n the getNetwork ID method, wrote unit test indepth edge cases in the program

### CODE REVIEW

File: pkg/openstack/loadbalancer_test.go [hunk 1]
```diff
@@ -534,3 +536,104 @@ func Test_getListenerProtocol(t *testing.T) {
...
+			wantErr: nil,
+		},
+		{
+			name:   "get network id from default config",
+			fields: fields{},
+			args: args{
+				service: &corev1.Service{},
+				svcConf: &serviceConfig{},
+			},
+			want:    "",
```
Comment (dulek) [inline pkg/openstack/loadbalancer_test.go:605]: Can you populate config to make this actually return value from it?

File: pkg/openstack/loadbalancer_test.go [hunk 2]
```diff
@@ -534,3 +536,104 @@ func Test_getListenerProtocol(t *testing.T) {
...
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			lbaas := tt.fields.lbaas
+
+			got, err := lbaas.getNetworkID(tt.args.service, tt.args.svcConf)
+			if (err != nil) != (tt.wantErr != nil) || (err != nil && err.Error() != tt.wantErr.Error()) {
+				t.Errorf("LbaasV2.getNetworkID() error = %v, wantErr %v", err, tt.wantErr)
+				return
+			}
+			if got != tt.want {
```
Comment (dulek) [inline pkg/openstack/loadbalancer_test.go:634]: Please use `assert.Equal` instead of a manual comparison.

File: pkg/openstack/loadbalancer_test.go [hunk 3]
```diff
@@ -534,3 +536,104 @@ func Test_getListenerProtocol(t *testing.T) {
...
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			lbaas := tt.fields.lbaas
+
+			got, err := lbaas.getNetworkID(tt.args.service, tt.args.svcConf)
+			if (err != nil) != (tt.wantErr != nil) || (err != nil && err.Error() != tt.wantErr.Error()) {
+				t.Errorf("LbaasV2.getNetworkID() error = %v, wantErr %v", err, tt.wantErr)
+				return
+			}
```
Comment (dulek) [inline pkg/openstack/loadbalancer_test.go:633]: Please make sure to use `assert.NoError()` and `assert.ErrorIs()` (or similar function that'd work).

File: pkg/openstack/loadbalancer_test.go [hunk 4]
```diff
@@ -534,3 +536,104 @@ func Test_getListenerProtocol(t *testing.T) {
...
+			opts: LoadBalancerOpts{
+				LBClasses: map[string]*LBClass{"lbclassKey": {NetworkID: "lb-class-network-id-1234"}},
+				NetworkID: "lb-class-networkId",
+			},
+		},
+	}
+
+	type fields struct {
+		lbaas LbaasV2
+	}
```
Comment (dulek) [inline pkg/openstack/loadbalancer_test.go:552]: Why do we need this `fields`? This feels out of place.

**Review by dulek — CHANGES_REQUESTED** [review]

**Comment by KingDaemonX** [comment]
@pierreprinetti can you review this
i recently made all the needed fix and want to be sure i am right on track

File: pkg/openstack/loadbalancer_test.go [hunk 5]
```diff
@@ -1543,3 +1543,99 @@ func Test_nodeAddressForLB(t *testing.T) {
...
+			},
+		},
+	}
+
+	type testArg struct {
+		service *corev1.Service
+	}
+	tests := []struct {
+		name    string
+		lbaas   LbaasV2
```
Comment (pierreprinetti) [inline pkg/openstack/loadbalancer_test.go:1566]: I suggest using a pointer here for consistency with the function receiver. (or you can use a bare value if you want, but then please define the variable `lbaas` as a bare value too line 1548)

```suggestion
		lbaas   *LbaasV2
```

File: pkg/openstack/loadbalancer_test.go [hunk 6]
```diff
@@ -1543,3 +1543,99 @@ func Test_nodeAddressForLB(t *testing.T) {
...
+
+			got, err := lbaas.getNetworkID(tt.arg.service, &serviceConfig{})
+
+			if tt.wantErr != "" {
+				assert.EqualError(t, err, tt.wantErr)
+			} else {
+				if assert.NoError(t, err) {
+					assert.Equal(t, tt.want, got)
+				}
+			}
```
Comment (pierreprinetti) [inline pkg/openstack/loadbalancer_test.go:1638]: Why is there a nested `if`?

In this `else` branch, `tt.wantErr` is the empty string. This means that we expect `err` to be nil, full stop.

```suggestion
			} else {
				assert.NoError(t, err)
			}

Comment (KingDaemonX) [reply pkg/openstack/loadbalancer_test.go:1638]: image

i tried following the suggested principle that assert if the got and expected are equal if there is no error :)

i made the fixed already and sent them in @pierreprinetti

Comment (pierreprinetti) [reply pkg/openstack/loadbalancer_test.go]: ah! So that is where this pattern is coming from, thanks.

Comment (pierreprinetti) [reply pkg/openstack/loadbalancer_test.go]: I don't like it.

**Review by pierreprinetti — APPROVED** [review]
Nice!
/lgtm
/assign jichenjc

File: pkg/openstack/loadbalancer_test.go (hunk 6 above)
Comment (KingDaemonX) [reply pkg/openstack/loadbalancer_test.go]: yeah, i fixed it already :-)

**Comment by pierreprinetti** [comment]
Oh this needs a rebase unfortunately @KingDaemonX

File: pkg/openstack/loadbalancer_test.go (hunk 6 above)
Comment (KingDaemonX) [reply pkg/openstack/loadbalancer_test.go]: apparently, i'm learning alot in the process 
thank you very much

**Comment by KingDaemonX** [comment]
image

it does not have conflict or needs rebase here, so i am confuse

**Comment by pierreprinetti** [comment]
> image
> 
> it does not have conflict or needs rebase here, so i am confuse

Refresh perhaps? Github also shows a warning here.

**Review by pierreprinetti — APPROVED** [review]

**Comment by KingDaemonX** [comment]
> > image
> > it does not have conflict or needs rebase here, so i am confuse
> 
> Refresh perhaps? Github also shows a warning here.

apparently there was a new merge to the master

**Comment by KingDaemonX** [comment]
@jichenjc rebased already and ready for review

File: pkg/openstack/loadbalancer_test.go [hunk 7]
```diff
@@ -1754,3 +1754,99 @@ func TestBuildBatchUpdateMemberOpts(t *testing.T) {
...
+			},
+			want:    "",
+			wantErr: "invalid loadbalancer class \"invalid-lb-class\"",
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			lbaas := tt.lbaas
+
+			got, err := lbaas.getNetworkID(tt.arg.service, &serviceConfig{})
```
Comment (pierreprinetti) [inline pkg/openstack/loadbalancer_test.go:1841]: just a nit: you don't need to use yet another variable

```suggestion
			got, err := tt.lbaas.getNetworkID(tt.arg.service, &serviceConfig{})
```

**Review by pierreprinetti — COMMENTED** [review]
/lgtm

File: pkg/openstack/loadbalancer_test.go (hunk 7 above)
Comment (KingDaemonX) [reply pkg/openstack/loadbalancer_test.go:1841]: I totally forgot, probably went back when I undo an error 
I had issues that I was fixing and undo some changes, so I think this sneaked in

**Comment by KingDaemonX** [comment]
@jichenjc are you available

**Review by pierreprinetti — APPROVED** [review]
/lgtm

**Comment by KingDaemonX** [comment]
@pierreprinetti how do i remove the `needs rebase`  tag 
the PR is rebased and still having the label

**Comment by pierreprinetti** [comment]
> @pierreprinetti how do i remove the `needs rebase` tag the PR is rebased and still having the label

The bot was on vacation apparently!

**Comment by KingDaemonX** [comment]
> > @pierreprinetti how do i remove the `needs rebase` tag the PR is rebased and still having the label
> 
> The bot was on vacation apparently!

😅 classic
